### ISSUE TITLE
Error linking Compose runtime APIs when building XC Framework - causes runtime crash on iOS

### ISSUE THREAD

**Issue by kirillzh** [JetBrains/compose-multiplatform#3611]
**Describe the bug**
When building XC Framework for KMP module that uses Compose runtime, getting a bunch of warnings `No function found for symbol...` for Compose runtime APIs. XC Framework does built successfully, when using on iOS, it results in runtime crash with the same error that was mentioned in the warning.

This issue specifically started happening after I enabled K/N cache by removing `kotlin.native.cacheKind=none` from my `gradle.properties`. 

```
Function 'LaunchedEffect' can not be called: No function found for symbol 'androidx.compose.runtime/LaunchedEffect|-576771147392862334[0]'
w: <missing declarations>: No function found for symbol 'app.cash.molecule/launchMolecule|-7305106088441339121[0]'
w: <missing declarations>: No function found for symbol 'androidx.compose.runtime/remember|1835138394250687683[0]'
w: <missing declarations>: No function found for symbol 'androidx.compose.runtime/remember|-2596345568877061727[0]'
w: <missing declarations>: No function found for symbol 'androidx.compose.runtime/LaunchedEffect|-576771147392862334[0]'
w: <missing declarations>: No function found for symbol 'androidx.compose.runtime/collectAsState|-5046421723490222442[0]'
w: <missing declarations>: No function found for symbol 'androidx.compose.runtime/remember|-1588588864557296532[0]'
w: <missing declarations>: No function found for symbol 'androidx.compose.runtime/LaunchedEffect|6204867515921259929[0]'
w: <missing declarations>: No function found for symbol 'androidx.compose.runtime/remember|6302736979574676244[0]'
w: <missing declarations>: No function found for symbol 'androidx.compose.runtime/collectAsState|-6899767239674529654[0]'
w: <missing declarations>: No function found for symbol 'androidx.compose.runtime/rememberCoroutineScope|5952073114146778194[0]'
w: <missing declarations>: No function found for symbol 'androidx.compose.runtime/produceState|-2115622692291411606[0]'
w: <missing declarations>: No function found for symbol 'androidx.compose.runtime/LaunchedEffect|-9057022742877438542[0]'
w: <missing declarations>: No function found for symbol 'androidx.compose.runtime/DisposableEffect|4541906961598546030[0]'
w: <missing declarations>: No function found for symbol 'androidx.compose.runtime/DisposableEffect|-9014090276099959147[0]'
w: <missing declarations>: No function found for symbol 'androidx.compose.runtime/remember|-2215966373931868872[0]'
```

**Affected platforms**
- iOS

**Versions**
- *Kotlin version*:  1.9.0
- *Compose Multiplatform version*:  1.5.0
- *OS version(s)*: macOS 13.5.1 
- *OS architecture (x86 or arm64)*:  arm64
- *JDK*: 17

**Comment by m-sasha** [comment]
Is this the same issue as https://github.com/JetBrains/compose-multiplatform/issues/3318 ?

**Comment by AlexeyTsvetkov** [comment]
@kirillzh I was not able to reproduce the issue. Could you share a reproducible sample or your Gradle configuration (the exact dependencies can be omitted) at least for a module, that causes the error?

> This issue specifically started happening after I enabled K/N cache by removing kotlin.native.cacheKind=none from my gradle.properties.

That's weird, because with Compose 1.5.0 and Kotlin 1.9.0 we disable caches ourselves, when `org.jetbrains.compose` Gradle plugin is applied in the same module as `org.jetbrains.kotlin.multiplatform`.

Also, could you try adding the following code as a workaround?
```
import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinNativeTarget
// ...
kotlin {
    targets.configureEach {
        if (this is KotlinNativeTarget) {
            compilations.configureEach {
                kotlinOptions.freeCompilerArgs += listOf("-Xlazy-ir-for-caches=disable")
            }
        }
    }
}
```

**Comment by kirillzh** [comment]
> Could you share a reproducible sample or your Gradle configuration (the exact dependencies can be omitted) at least for a module, that causes the error?

This is happening in a module that builds XC Framework for iOS, when running `:assembleSharedDebugXCFramework` task specifically. It's part of a big project which I can't share but here's some more info:
- A bunch of modules in the project use Compose KMP plugin and expose Compose runtime deps in ABI
- Many KMP libraries used: Kmp NativeCoroutines, Serialization, Molecule, Turbine, Kermit, Ktor, SqlDelight, Wire, etc.
- Here's how the XC Framework module is roughly configured:
```kotlin
plugins {
  // plugins: kotlin.multiplatform, crashkios.bugsnag
}

kotlin {
  // kmp ios source sets and targets configuration
  val runtimeDependencies = listOf(
    // modules (many of them use and expose Compose runtime APIs), libs
  )

  sourceSets {
    commonMain {
      dependencies {
        runtimeDependencies.onEach { dep ->
          api(dep)
        }
      }
    }
  }

  val exposedDependencies = listOf(
    // modules (many of them use and expose Compose runtime APIs), libs
  )

  val xcFramework = XCFramework(xcFrameworkName = "Shared")

  // optimalTargetsForIOS picks which ios targets to build for based on host architecture
  optimalTargetsForIOS(project).forEach { 
    it.binaries.framework(DEBUG) {
      baseName = "Shared"

      // Disable iOS bitcode.
      embedBitcode(DISABLE)

      xcFramework.add(this)
      exposedDependencies.onEach { dep ->
        export(dep)
      }
    }

    it.linkSqlite()
  }
}

/**
 * Manually linking Sqlite3 binaries.
 *
 * https://github.com/touchlab/SQLiter/issues/88
 * https://github.com/touchlab/SQLiter/issues/77
 */
fun KotlinNativeTarget.linkSqlite() {
  binaries.configureEach {
    linkerOpts += "-lsqlite3"
  }
}
```

**Comment by kirillzh** [comment]
@AlexeyTsvetkov, adding `-Xlazy-ir-for-caches=disable` compiler flag for native targets seems to the fix the build, thanks! 

Though, now that K/N cache is enabled, I'm not seeing too much improvement in build times (at most 10 seconds with incremental changes, which is still great). What does `-Xlazy-ir-for-caches=disable` flag do? Is it expected that passing it results in worse caching capability?

**Comment by AlexeyTsvetkov** [comment]
> Though, now that K/N cache is enabled, I'm not seeing too much improvement in build times (at most 10 seconds with incremental changes, which is still great). What does -Xlazy-ir-for-caches=disable flag do? Is it expected that passing it results in worse caching capability?

K/N team estimates that the effect of `-Xlazy-ir-for-caches=disable` should be in range of 10-20% for "linkKotlin*" tasks https://youtrack.jetbrains.com/issue/KT-45823/Implement-Lazy-IR-linkage-for-cached-dependencies

AFAIK K/N compiler caches are caching some part of the transformation pipeline for external libraries only. Your code still needs to be compiled into backend IR first (this is what "compileKotlin*" tasks do), then your IR needs to be linked and lowered with external libraries IR into LLVM IR,
then LLVM IR gets optimized and only then it is transformed into machine code. Most of the pipeline is still not incremental. However, the K/N team continues to work on the compilation pipeline improvements. According to [the roadmap issue](https://youtrack.jetbrains.com/issue/KT-42294/Improve-Kotlin-Native-compilation-time), alpha version of the parallel incremental compilation for the second stage of the compilation should become available in one of the next releases. Feel free to follow https://youtrack.jetbrains.com/issue/KT-42294 for the updates

Anyway there is no way to use K/N caches without `-Xlazy-ir-for-caches=disable` with Compose and Kotlin 1.9.0. Either you turn them off completely or set `-Xlazy-ir-for-caches=disable` or don't use Compose. Caching issues are expected to be fixed in Kotlin 1.9.20.

**Comment by okushnikov** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/CMP-3611">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.
